src/conv/invokers/impl_gemm_dynamic.cpp

```diff
@@ -239,11 +239,13 @@ float CallImplGemmDynamicBackwardData(const miopen::Handle& handle,
         opArgs[26]             = OpKernelArg(_y_dot_slice);
         opArgs[27]             = OpKernelArg(_x_dot_slice);
         if(is_gemm_not_empty)
+        {
```

Comment:
This function performs many computations, thus wasting time. The most of those can be (and therefore must be) moved into the Invoker Factory. An Invoker instance will hold copies of the values computed _once_ by the its Factory.

Reply:
Hi artem,
I don't quite understand your comments, MakeImplGemmDynamicBackwardDataInvokerFactory is called by solver when GetSolution(), and for each convolution, we need to pass specific kernArg into kernels, and for bwd, we need pre-computation to determine if we need launch several kernels or not.

Reply:
The library keeps instances of Invokers in the cache, -- an instance per each convolution (problem config). That is why Invoker can hold parameters that are invariant for a problem config. For example, `n` (and, generally, the most of the parameters computed in this function, that depend only on problem config). This way, Invoker is not required to compute these parameters (or take these from ProblemDescription) each time when we want to run a kernel. This is one of the basic Invoker concepts.

The Factory should compute these parameters and pass to Invoker by-value (it is a lambda, so it it will hold copies). When library needs Invoker for some convolution (problem config), it will find it in the cache. Invoker will "read" invariant values from itself ans pass as parameters to the kernel. The only parameters that we need to pass to Invoker each time are, basically, input & output buffer addresses.

Reply:
@carlushuang ,what do you think of this comments, since it's initially created by you.

Reply:
@atamazov , thanks, actually, our asm-igemm kernels need many kernels arguments, like buffer pointer, convolution parameters(like n, c, k, hi, wi, pad, stride, etc, also other config parameters as well), I'm quite don't understand how to make all these parameters avalable in 'invoker', but not get and compute from 'ConvolutionContext& ctx', is there example code existed in MIOpen now?

Reply:
@jane-zxy  @atamazov  below is a tentative change for this invoker. This move some computation outside lambda, so within lambda now there is only 4 lines of computation, since bwd need multiple launch
```
InvokerFactory MakeImplGemmDynamicBackwardDataInvokerFactory(const ConvolutionContext& ctx)
{
    const auto& conv_problem = ctx.conv_problem;
    // clang-format off
    int hi          = conv_problem.GetOutHeight();
    int wi          = conv_problem.GetOutWidth();
    int n           = conv_problem.GetInBatchSize();
    int k           = conv_problem.GetInChannels();
    int c           = conv_problem.GetOutChannels();
    int ho          = conv_problem.GetInHeight();
    int wo          = conv_problem.GetInWidth();
    int stride_h    = conv_problem.GetInHeight() > 1 ? conv_problem.GetKernelStrideH() : 1;
    int stride_w    = conv_problem.GetInWidth() > 1 ? conv_problem.GetKernelStrideW() : 1;
    int dilation_h  = conv_problem.GetWeightsHeight() > 1? conv_problem.GetDilationH() : 1;
    int dilation_w  = conv_problem.GetWeightsWidth() > 1? conv_problem.GetDilationW() : 1;
    int pad_h       = conv_problem.GetPadH();
    int pad_w       = conv_problem.GetPadW();
    int y           = conv_problem.GetWeightsHeight();
    int x           = conv_problem.GetWeightsWidth();

    int gcd_stride_dilation_h = gcd(stride_h, dilation_h);
    int gcd_stride_dilation_w = gcd(stride_w, dilation_w);
    int y_tilda     = stride_h / gcd_stride_dilation_h;
    int x_tilda     = stride_w / gcd_stride_dilation_w;

    int y_dot = (y +  y_tilda - 1) / y_tilda;
    int x_dot = (x +  x_tilda - 1) / x_tilda;

    int h_tilda     = ho + (dilation_h * (y - 1) + stride_h - 1) / stride_h;
    int w_tilda     = wo + (dilation_w * (x - 1) + stride_w - 1) / stride_w;

    int h_tilda_left = std::max(0, pad_h - dilation_h * (y_tilda - 1)) / stride_h;
    int w_tilda_left = std::max(0, pad_w - dilation_w * (x_tilda - 1)) / stride_w;

    int h_tilda_right = std::min(h_tilda, (pad_h + hi - 1 + stride_h - 1) / stride_h + 1);
    int w_tilda_right = std::min(w_tilda, (pad_w + wi - 1 + stride_w - 1) / stride_w + 1);

    int h_tilda_slice = h_tilda_right - h_tilda_left;
    int w_tilda_slice = w_tilda_right - w_tilda_left;

    int num_of_gemms  = x_tilda * y_tilda;

    //int dtile_iy    = 0;
    //int dtile_ix    = 0;
    int dtile_dy    = dilation_h / gcd_stride_dilation_h;
    int dtile_dx    = dilation_w / gcd_stride_dilation_w;
    int dtile_y     = y_tilda;
    int dtile_x     = x_tilda;
    int dtile_h     = h_tilda;
    int dtile_w     = w_tilda;
    //int dslice_y    = 0;
    //int dslice_x    = 0;
    int dslice_h    = h_tilda_slice;
    int dslice_w    = w_tilda_slice;
    int dslice_h_left   = h_tilda_left;
    int dslice_w_left   = w_tilda_left;
    int __pack0     = 0;
    // clang-format on

    return [=](const std::vector<Kernel>& kernels) {
        const auto kernel = kernels[0];
        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) {
            decltype(auto) data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
            const auto& tensors     = data_ctx.tensors;

            for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
            {
                int dtile_iy_gid          = gemm_id / x_tilda;
                int dtile_ix_gid          = gemm_id % x_tilda;
                int y_dot_slice_gid       = (dtile_iy_gid + 1) * y_dot <= y ? y_dot : y % y_dot;
                int x_dot_slice_gid       = (dtile_ix_gid + 1) * x_dot <= x ? x_dot : x % x_dot;
                int gemm_k_gid            = k * y_dot_slice_gid * x_dot_slice_gid;
                bool is_gemm_not_empty = gemm_k_gid > 0;

                if(is_gemm_not_empty)
                {
                    handle.Run(kernel)(
                        tensors.out,
                        tensors.w,
                        tensors.in,
                        hi,
                        wi,
                        n,
                        k,
                        c,
                        ho,
                        wo,
                        stride_h,
                        stride_w,
                        dilation_h,
                        dilation_w,
                        pad_h,
                        pad_w,
                        y,
                        x,
                        dtile_iy_gid,
                        dtile_ix_gid,
                        dtile_dy,
                        dtile_dx,
                        dtile_y,
                        dtile_x,
                        dtile_h,
                        dtile_w,
                        y_dot_slice_gid,
                        x_dot_slice_gid,
                        dslice_h,
                        dslice_w,
                        dslice_h_left,
                        dslice_w_left,
                        __pack0);
                }
            }
        };
    };
}
```

Reply:
Yes, I think this is correct implementation of the approach. Except `__pack0` which is reserved identifier and must be renamed ;)

Reply:
Please also try to remove `//clang-format` stuff, perhaps the resulting formatting will be acceptable.

Reply:
Also I recommend moving the computational contents of the loop to the factory, and using arrays for storing results. In Factory:
```C++
//define arrays and then...
for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
{
    dtile_iy_gid[gemm_id]    = gemm_id / x_tilda;
    dtile_ix_gid[gemm_id]    = gemm_id % x_tilda;
    y_dot_slice_gid[gemm_id] = (dtile_iy_gid[gemm_id] + 1) * y_dot <= y ? y_dot : y % y_dot;
    x_dot_slice_gid[gemm_id] = (dtile_ix_gid[gemm_id] + 1) * x_dot <= x ? x_dot : x % x_dot;
    const int gemm_k_gid     = k * y_dot_slice_gid[gemm_id] * x_dot_slice_gid[gemm_id];
    is_gemm_not_empty[gemm_id] = gemm_k_gid > 0;
}
```
In Invoker:
```C++
for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
{
    if(!is_gemm_not_empty[gemm_id])
    	continue;
     handle.Run(kernel)(
            tensors.out,
            tensors.w,
            tensors.in,
            hi,
            wi,
            n,
            k,
            c,
            ho,
            wo,
            stride_h,
            stride_w,
            dilation_h,
            dilation_w,
            pad_h,
            pad_w,
            y,
            x,
            dtile_iy_gid[gemm_id],
            dtile_ix_gid[gemm_id],
            dtile_dy,
            dtile_dx,
            dtile_y,
            dtile_x,
            dtile_h,
            dtile_w,
            y_dot_slice_gid[gemm_id],
            x_dot_slice_gid[gemm_id],
            dslice_h,
            dslice_w,
            dslice_h_left,
            dslice_w_left,
            __pack0);
}

```

Reply:
OK, @jane-zxy  let's follow the comment and update into https://github.com/ROCmSoftwarePlatform/MIOpen/pull/508

Reply:
[Resolved] Let's keep this in non-hidden state to allow hyperlinks to this thread.